process/connections.go

```diff
@@ -34,6 +34,26 @@ func (m *CollectorConnections) IterateDNS(addr *Addr, cb func(i, total int, entr
 }
 
 // GetDNSNames returns all the DNS entries
-func (m *CollectorConnections) GetDNSNames() []string {
+func (m *CollectorConnections) GetDNSNames() ([]string, error) {
 	return getDNSNames(m.EncodedDNS)
 }
+
+// GetDNSV2 returns the DNS entries for the given addr.
+// The first argument returned is the first DNS entry followed by any additional name resolutions.  Most IPs will
+// have a single resolution so this dual format allows us to avoid allocations for the common case.  If there are
+// multiple name resolutions, there is no implied priority between the dual values
+// the return values are indexes into the dns database (GetDNSDatabase)
+
+func (m *CollectorConnections) GetDNSV2(addr *Addr) (int32, []int32, error) {
+	return GetDNSV2(m.EncodedDnsLookups, addr.Ip)
+}
+
+// IterateDNSV2 iterates over all of the DNS entries for the given addr, invoking the provided callback for each one
+func (m *CollectorConnections) IterateDNSV2(addr *Addr, cb func(i, total int, entry int32) bool) {
```

Comment:
Can we have just one `CollectorConnections.IterateDNS` that will figure out the `dnsVersion` and call the appropriate `IterateDNS`?

Reply:
The interface is going to have to change _somewhere_.  Since we don't have the buffer, we can't return the string, just the offset into the string.  so if we _provide_ the buffer, the interface changes.
Furthermore, if we return the string, this will cause an allocation each time, whereas there was a specific request to not do an allocation for each string.  @kevinconaway

Reply:
>Since we don't have the buffer, we can't return the string

I am not sure I follow. Earlier, we had both the offset and the string in `EncodedDNS` and was sending a slice of the `EncodedDNS` to the callback, instead of allocating a string.

Can't we do something similar here and send a slice from `EncodedDomainDatabase` (instead of `EncodedDNS`) without allocating any string?

Reply:
the iterate function doesn't have the actual byte slice.

Reply:
It currently doesn't but can we pass it the slice?

Currently, we have the following:
```
func (m *CollectorConnections) IterateDNS(addr *Addr, cb func(i, total int, entry string) bool) {
	IterateDNS(m.EncodedDNS, addr.Ip, cb)
}
```
Can't it be modified to do the following?
```
func (m *CollectorConnections) IterateDNS(addr *Addr, cb func(i, total int, entry string) bool) {
        if len(m.EncodedDNS) > 0 {
	      IterateDNS(m.EncodedDNS, addr.Ip, cb)
        } else {
             IterateDNSV2(m.EncodedDnsLookups, m.EncodedDomainDatabase, addr.Ip, cb)
        }
}
```

Reply:
We can, but then the interface is still changing... unless I'm missing _which_ interface you want to be consistent.  The _encoder_ interface will still be different, which is the interface defined in this pr/repo.

Reply:
Right, the encoder has to change. But backend apps (`dd-go` repo) only to decode (encoding in done some places for testing).

Here is how the backend iterates through the list of domains for a given IP:

- https://github.com/DataDog/dd-go/blob/prod/process/apps/network-logger/model/host.go#L664
- https://github.com/DataDog/dd-go/blob/prod/networks/model/host.go#L353

Reply:
`func (m *CollectorConnections) IterateDNS(addr *Addr, cb func(i, total int, entry string) bool) {`

Provides a callback (cb). If we make the change you're asking for, then IIUC every call to `cb` will require an allocation (converting the appropriate section of the byte slice into a string).  There was a specfiic requirement _not_ to do that.

Reply:
> And you still have the problem that each call to the iterator callback function will result in a string allocation

Can you elaborate on that? Why will calling the iterator callback function with `EncodedDomainDatabase[start:finish]` will result in string allocation when the current approach of calling the function with `EncodedDNS[start:finish]` does not?

Reply:
My understanding is that calling `string(slice[start:finish])` causes an allocation @kevinconaway

Reply:
This is what we currently pass to the callback:
https://github.com/DataDog/agent-payload/blob/master/process/dns_v1.go#L353
```
if !cb(j, int(nameCount), nameBuffer[start:start+int(nameLength)]) {
     return
}
```
In v1,  `nameBuffer` is a part of the byte array `EncodedDNS`. If it becomes part of the byte array `EncodedDomainDatabase` in `v2`, why will it require string allocation when it doesn't in v1? Is there a difference that I am missing?